scipy/linalg/src/_common_array_utils.h

```diff
@@ -0,0 +1,222 @@
+void BLAS_FUNC(dgecon)(char* norm, CBLAS_INT* n, double* a,      CBLAS_INT* lda, double* anorm, double* rcond, double* work,      CBLAS_INT* iwork, CBLAS_INT* info);
+void BLAS_FUNC(cgecon)(char* norm, CBLAS_INT* n, npy_cfloat* a,  CBLAS_INT* lda, float* anorm,  float* rcond,  npy_cfloat* work,  float* rwork,     CBLAS_INT* info);
+void BLAS_FUNC(zgecon)(char* norm, CBLAS_INT* n, npy_cdouble* a, CBLAS_INT* lda, double* anorm, double* rcond, npy_cdouble* work, double* rwork,    CBLAS_INT* info);
+
+
+/* ?TRTRI */
+void BLAS_FUNC(strtri)(char * uplo, char *diag, CBLAS_INT *n, float* a, CBLAS_INT *lda, CBLAS_INT *info);
+void BLAS_FUNC(dtrtri)(char * uplo, char *diag, CBLAS_INT *n, double* a, CBLAS_INT *lda, CBLAS_INT *info);
+void BLAS_FUNC(ctrtri)(char * uplo, char *diag, CBLAS_INT *n, npy_cfloat* a, CBLAS_INT *lda, CBLAS_INT *info);
+void BLAS_FUNC(ztrtri)(char * uplo, char *diag, CBLAS_INT *n, npy_cdouble* a, CBLAS_INT *lda, CBLAS_INT *info);
+
+
+} // extern "C"
+
+
+/*
+ * Generate type overloads, to map from C array types (float, double, npy_cfloat, npy_cdouble)
+ * to LAPACK prefixes, "sdcz".
+ */
+#define GEN_GETRF(PREFIX, TYPE) \
+inline void \
+getrf(CBLAS_INT *m, CBLAS_INT *n, TYPE *a, CBLAS_INT *lda, CBLAS_INT *ipiv, CBLAS_INT *info) \
+{ \
+    BLAS_FUNC(PREFIX ## getrf)(m, n, a, lda, ipiv, info); \
+};
+
+GEN_GETRF(s,float)
+GEN_GETRF(d,double)
+GEN_GETRF(c,npy_cfloat)
+GEN_GETRF(z,npy_cdouble)
+
+#define GEN_GETRI(PREFIX, TYPE) \
+inline void \
+getri(CBLAS_INT *n, TYPE *a, CBLAS_INT *lda, CBLAS_INT *ipiv, TYPE *work, CBLAS_INT *lwork, CBLAS_INT *info) \
+{ \
+    BLAS_FUNC(PREFIX ## getri)(n, a, lda, ipiv, work, lwork, info); \
+};
+
+GEN_GETRI(s,float)
+GEN_GETRI(d,double)
+GEN_GETRI(c,npy_cfloat)
+GEN_GETRI(z,npy_cdouble)
+
+
+
+// NB: iwork for real arrays or rwork for complex arrays
+#define GEN_GECON(PREFIX, CTYPE, RTYPE, WTYPE) \
+inline void \
+gecon(char* norm, CBLAS_INT* n, CTYPE* a, CBLAS_INT* lda, RTYPE* anorm, RTYPE* rcond, CTYPE* work, void *irwork, CBLAS_INT* info) \
+{ \
+    BLAS_FUNC(PREFIX ## gecon)(norm, n, a, lda, anorm, rcond, work, (WTYPE *)irwork, info); \
+};
+
+GEN_GECON(s, float, float, CBLAS_INT)
+GEN_GECON(d, double, double, CBLAS_INT)
+GEN_GECON(c, npy_cfloat, float, float)
+GEN_GECON(z, npy_cdouble, double, double)
+
+
+#define GEN_TRTRI(PREFIX, TYPE) \
+inline void \
+trtri(char* uplo, char *diag, CBLAS_INT* n, TYPE* a, CBLAS_INT* lda, CBLAS_INT* info) \
+{ \
+    BLAS_FUNC(PREFIX ## trtri)(uplo, diag, n, a, lda, info); \
+};
+
+
+GEN_TRTRI(s, float)
+GEN_TRTRI(d, double)
+GEN_TRTRI(c, npy_cfloat)
+GEN_TRTRI(z, npy_cdouble)
+
+
+
+/***************************
+ ***  Structure detection
+ ***************************/
+
+
+bool is_nzero(float value) {return value != 0.0f;}
+bool is_nzero(double value) {return value != 0.0;}
+bool is_nzero(npy_cfloat value) {return (npy_crealf(value) != 0.0f) && (npy_cimagf(value) != 0.0f);}
```

Comment:
These `&&` needs to be `||`

Reply:
Grrr. Thanks! 
I think `npy_cfloat` and `npy_double` arithmetics has crossed my annoyance threshold. Will convert to using native types. A couple of `reinterpret_cast`s, and all of these are gone.

Reply:
This is done now BTW.